engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java

```diff
@@ -870,6 +873,21 @@ private void execute() throws YarnException, IOException
       }
 
       numRequestedContainers += containerRequests.size() - removedContainerRequests.size();
+
```

Comment:
An alternative (more modular/OO) implementation could be to move this logic into the existing NMCallbackHandler class. Save the reference to NMCallbackHandler in StreamingAppMasterService (or get it from nmClient.getCallbackHandler) and move pendingContainersToKill in there. When NMCallbackHandler.onContainerStopped(ContainerId) is called remove the containerId from pendingContainersToKill and then only examine that list for containers that have been pending there for a long time.